### PR TITLE
FIRM variable importance

### PR DESCRIPTION
**Pull request by hideaki**
# Description
FIRM variable importance support.

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/build_lm.R [hunk 1]
```diff
@@ -72,17 +72,29 @@ partial_dependence.lm_exploratory <- function(fit, target, vars = colnames(data)
     c("preds" = mean(x))
   }
 
+  points <- list()
```
Comment (hideaki) [inline R/build_lm.R:75]: Add comment.

File: R/partial_dependence.R [hunk 2]
```diff
@@ -81,9 +90,10 @@ handle_partial_dependence <- function(x) {
     ret <- ret %>% dplyr::rename(Predicted=!!rlang::sym(target_col)) # Rename target column to Predicted to make comparison with Actual.
   }
   else if (!is.null(x$partial_binning)) { # For ranger/rpart, we calculate binning data at model building. Maybe we should do the same for glm/lm.
+    # Note that this case consists of only regression and binary classification, and not multiclass classification.
     actual_ret <- x$partial_binning
     ret <- actual_ret %>% dplyr::bind_rows(ret) # So that PDP is drawn over the binning, the bind_row order needs to be this way.
-    if (!is.null(ret[[target_col]])) {
+    if (length(target_col) == 1 && target_col != "TRUE" && !is.null(ret[[target_col]])) { # Regression case
```
Comment (kei51e) [inline R/partial_dependence.R:96]: Add description when target_col == "TRUE" case.

File: R/randomForest_tidiers.R [hunk 3]
```diff
@@ -2082,6 +2082,60 @@ importance_ranger <- function(model) {
   imp_df
 }
 
+# Standard deviation with weight by how many rows have the same value.
+# Using it to calculate FIRM while the PDP data has only single row even when multiple quantile-based grid points have a same value.
+sd_with_weight <- function(v, w) {
```
Comment (kei51e) [inline R/randomForest_tidiers.R:2087]: Move to partial_dependence.R.

File: R/randomForest_tidiers.R [hunk 4]
```diff
@@ -2082,6 +2082,60 @@ importance_ranger <- function(model) {
...
+  imps <- purrr::map(vectors, function(v) {
+    # Get FIRM value based on whether the variable is categorical or not.
+    if (class[[1]] == "numeric") {
+      sd_with_weight(v, weight)
+    }
+    else {
+      (max(v, na.rm=TRUE) - min(v, na.rm=TRUE))/4
+    }
+  })
+  imps <- purrr::flatten_dbl(imps)
```
Comment (kei51e) [inline R/randomForest_tidiers.R:2102]: Add comment. What is output?

File: R/randomForest_tidiers.R [hunk 5]
```diff
@@ -2082,6 +2082,60 @@ importance_ranger <- function(model) {
...
+  ret <- mean(imps, na.rm=TRUE)
```
Comment (kei51e) [inline R/randomForest_tidiers.R:2103]: Why we need mean?

File: R/randomForest_tidiers.R [hunk 6]
```diff
@@ -2082,6 +2082,60 @@ importance_ranger <- function(model) {
...
+
+# Caluculate FIRM variable importance.
+# References:
+#   https://arxiv.org/abs/1805.04755
+#   https://arxiv.org/abs/1904.03959
+importance_firm <- function(pdp_data, target, vars) {
+  points <- attr(pdp_data, "points")
+  # Replace the grid values with the type of the column, e.g. "numeric", or "character".
+  names(vars) <- NULL # Clean names, since it messes up following mutate step.
+  imp_df <- pdp_data %>% dplyr::mutate(across(!!vars, ~ifelse(is.na(.x), NA, class(.x))))
```
Comment (kei51e) [inline R/randomForest_tidiers.R:2115]: NA ->  as.character(NA)

File: tests/testthat/test_build_lm_2.R [hunk 7]
```diff
@@ -62,7 +87,23 @@ test_that("build_lm.fast (logistic regression(numeric)) evaluate training and te
...
+
+  ret <- model_df %>% prediction_binary(data="training_and_test", threshold = 0.5, pretty.name=TRUE)
+  ret <- model_df %>% prediction_binary(data="training_and_test", threshold = 0.5)
+  test_ret <- ret %>% filter(is_test_data==TRUE)
+  # expect_equal(nrow(test_ret), 1480) # Not very stable for some reason. Will revisit
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  # expect_equal(nrow(train_ret), 3454) # Not very stable for some reason. Will revisit
+  ret <- model_df %>% evaluate_binary_training_and_test("is delayed", pretty.name=TRUE, threshold=0.1)
+  expect_gt(ret$Recall[[1]], 0.8) # Expect Recall to be ok since threshold is as low as 0.1.
+  expect_equal(nrow(ret), 2) # 2 for train and test
```
Comment (kei51e) [inline tests/testthat/test_build_lm_2.R:102]: check the importance.

File: tests/testthat/test_build_lm_3.R [hunk 8]
```diff
@@ -21,13 +21,33 @@ if (!testdata_filename %in% list.files(testdata_dir)) {
...
+  # expect_equal(nrow(test_ret), 1475) # Not very stable for some reason. Will revisit.
+  expect_lt(nrow(test_ret), 1500)
+  expect_gt(nrow(test_ret), 1400)
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  # expect_equal(nrow(train_ret), 3444) # Not very stable for some reason. Will revisit.
+  expect_lt(nrow(train_ret), 3500)
+  expect_gt(nrow(train_ret), 3400)
+  ret <- model_df %>% evaluate_lm_training_and_test(pretty.name=TRUE)
+  expect_equal(nrow(ret %>% filter(`CAR RIER`=="DUMMY")), 1) # Row for the group with error.
+  expect_equal(nrow(ret %>% filter(`CAR RIER`=="DUMMY" & !is.na(Note))), 1) # Row for the group with error should have message in Note column.
```
Comment (kei51e) [inline tests/testthat/test_build_lm_3.R:46]: check the importance.

File: tests/testthat/test_build_lm_3.R [hunk 9]
```diff
@@ -47,7 +67,29 @@ test_that("build_lm.fast (linear regression) evaluate training and test", {
...
+  expect_lt(nrow(test_ret), 1500)
+  expect_gt(nrow(test_ret), 1400)
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  # expect_equal(nrow(train_ret), 3454) # Not very stable for some reason. Will revisit
+  expect_lt(nrow(train_ret), 3500)
+  expect_gt(nrow(train_ret), 3400)
+  ret <- model_df %>% evaluate_binary_training_and_test("is delayed", pretty.name=TRUE)
+  expect_equal(nrow(ret %>% filter(`CAR RIER`=="DUMMY")), 1) # Row for the group with error.
+  expect_equal(nrow(ret %>% filter(`CAR RIER`=="DUMMY" & !is.na(Note))), 1) # Row for the group with error should have message in Note column. 
+  # expect_equal(nrow(ret), 2) # 2 for train and test. Fails with group_by
```
Comment (kei51e) [inline tests/testthat/test_build_lm_3.R:88]: check the importance.

File: tests/testthat/test_randomForest_tidiers_3.R [hunk 10]
```diff
@@ -21,7 +21,45 @@ if (!testdata_filename %in% list.files(testdata_dir)) {
...
+
+  model_df <- flight %>%
+                calc_feature_imp(`FL NUM`, `DIS TANCE`, `DEP TIME`, test_rate = 0, pd_with_bin_means = TRUE)
+  ret <- model_df %>% prediction(data="training_and_test")
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  #expect_equal(nrow(train_ret), 4894) # Less than 5000 because of NAs in the target variable. Linux seems to have different result. Work around for now.
+  expect_lt(nrow(train_ret), 5000)
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 1) # 1 for train
```
Comment (kei51e) [inline tests/testthat/test_randomForest_tidiers_3.R:59]: check the importance.

File: tests/testthat/test_randomForest_tidiers_3.R [hunk 11]
```diff
@@ -63,7 +101,50 @@ test_that("calc_feature_imp(regression) evaluate training and test", {
...
+  expect_gt(nrow(train_ret), 3400)
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 2) # 2 for train and test
+
+  ret <- rf_evaluation_training_and_test(model_df, type = "evaluation_by_class")
+  expect_equal(nrow(ret), 4) # 4 for train/test times TRUE/FALSE
+
+  ret <- rf_evaluation_training_and_test(model_df, type = "conf_mat")
+  expect_equal(nrow(ret), 8) # 8 for train/test times predicted TRUE/FALSE times actual TRUE/FALSE
```
Comment (kei51e) [inline tests/testthat/test_randomForest_tidiers_3.R:131]: check the importance.

File: tests/testthat/test_randomForest_tidiers_3.R [hunk 12]
```diff
@@ -223,7 +304,39 @@ test_that("calc_feature_map(binary) evaluate training and test with SMOTE", {
...
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 2) # 2 for train and test
+
+  ret <- rf_evaluation_training_and_test(model_df, type = "evaluation_by_class")
+  ret <- rf_evaluation_training_and_test(model_df, type = "conf_mat")
+
+  model_df <- flight %>%
+                calc_feature_imp(`ORI GIN`, `DIS TANCE`, `DEP TIME`, test_rate = 0, pd_with_bin_means = TRUE)
+  ret <- model_df %>% prediction(data="training_and_test")
```
Comment (kei51e) [inline tests/testthat/test_randomForest_tidiers_3.R:330]: check the importance.

File: tests/testthat/test_rpart_2.R [hunk 13]
```diff
@@ -59,7 +60,87 @@ test_that("exp_rpart(regression) evaluate training and test", {
...
+
+  ret <- model_df %>% rf_evaluation_training_and_test()
+  expect_equal(nrow(ret), 2) # 2 for train and test
+  expect_equal(is.na(ret$auc), c(F,F)) # 2 for train and test
+
+  # Training only case
+  model_df <- flight %>% dplyr::mutate(is_delayed = as.logical(`is delayed`)) %>%
+                exp_rpart(is_delayed, `DIS TANCE`, `DEP TIME`, test_rate = 0)
+  ret <- model_df %>% prediction(data="training_and_test")
+  train_ret <- ret %>% filter(is_test_data==FALSE)
```
Comment (kei51e) [inline tests/testthat/test_rpart_2.R:136]: check the importance.

File: tests/testthat/test_rpart_2.R [hunk 14]
```diff
@@ -232,7 +313,38 @@ test_that("exp_rpart(binary) evaluate training and test with SMOTE", {
...
+  expect_lt(nrow(train_ret), 3500)
+  expect_gt(nrow(train_ret), 3400)
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 2) # 2 for train and test
+
+  # Training only case
+  model_df <- flight %>%
+                exp_rpart(`ORI GIN`, `DIS TANCE`, `DEP TIME`, test_rate = 0)
+  ret <- model_df %>% prediction(data="training_and_test")
```
Comment (kei51e) [inline tests/testthat/test_rpart_2.R:339]: check the importance.

File: tests/testthat/test_xgboost_1.R [hunk 15]
```diff
@@ -99,7 +146,62 @@ test_that("exp_xgboost - regression - evaluate training and test with locale con
...
+  ret <- model_df %>% prediction(data="training_and_test")
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  #expect_equal(nrow(train_ret), 4905) # Less than 5000 because of NAs in the target variable. Linux seems to have different result. Work around for now.
+  expect_lt(nrow(train_ret), 5000)
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 1) # 1 for train
+
+  # ret <- rf_evaluation_training_and_test(model_df, type = "evaluation_by_class") # Not implemented.
+  ret <- rf_evaluation_training_and_test(model_df, type = "conf_mat")
```
Comment (kei51e) [inline tests/testthat/test_xgboost_1.R:201]: check the importance.

File: tests/testthat/test_xgboost_1.R [hunk 16]
```diff
@@ -286,7 +388,40 @@ test_that("exp_xgboost(binary) evaluate training and test with SMOTE", {
...
+
+  model_df <- flight %>%
+                exp_xgboost(`ORI GIN`, `DIS TANCE`, `DEP TIME`, test_rate = 0, pd_with_bin_means = TRUE)
+  ret <- model_df %>% prediction(data="training_and_test")
+  train_ret <- ret %>% filter(is_test_data==FALSE)
+  #expect_equal(nrow(train_ret), 4944) # Linux seems to have different result. Work around for now.
+  expect_lt(nrow(train_ret), 5000)
+
+  ret <- rf_evaluation_training_and_test(model_df)
+  expect_equal(nrow(ret), 1) # 1 for train
```
Comment (kei51e) [inline tests/testthat/test_xgboost_1.R:421]: check the importance.

File: R/partial_dependence.R (hunk 2 above)
Comment (hideaki) [reply R/partial_dependence.R:96]: Done.

File: R/randomForest_tidiers.R (hunk 4 above)
Comment (hideaki) [reply R/randomForest_tidiers.R:2102]: Done.

File: R/randomForest_tidiers.R (hunk 5 above)
Comment (hideaki) [reply R/randomForest_tidiers.R:2103]: Done.

File: R/randomForest_tidiers.R (hunk 6 above)
Comment (hideaki) [reply R/randomForest_tidiers.R:2115]: I'm using NA_character_.

File: R/randomForest_tidiers.R (hunk 3 above)
Comment (hideaki) [reply R/randomForest_tidiers.R:2087]: Moved the code.

File: tests/testthat/test_build_lm_2.R (hunk 7 above)
Comment (hideaki) [reply tests/testthat/test_build_lm_2.R:102]: done.

File: tests/testthat/test_build_lm_3.R (hunk 8 above)
Comment (hideaki) [reply tests/testthat/test_build_lm_3.R:49]: Done.

File: tests/testthat/test_build_lm_3.R (hunk 9 above)
Comment (hideaki) [reply tests/testthat/test_build_lm_3.R:94]: Done.

File: tests/testthat/test_randomForest_tidiers_3.R (hunk 10 above)
Comment (hideaki) [reply tests/testthat/test_randomForest_tidiers_3.R:63]: Done.

File: tests/testthat/test_randomForest_tidiers_3.R (hunk 11 above)
Comment (hideaki) [reply tests/testthat/test_randomForest_tidiers_3.R:139]: Done.

File: tests/testthat/test_randomForest_tidiers_3.R (hunk 12 above)
Comment (hideaki) [reply tests/testthat/test_randomForest_tidiers_3.R:342]: Done.

File: tests/testthat/test_rpart_2.R (hunk 13 above)
Comment (hideaki) [reply tests/testthat/test_rpart_2.R:141]: Done.

File: tests/testthat/test_rpart_2.R (hunk 14 above)
Comment (hideaki) [reply tests/testthat/test_rpart_2.R:348]: Done.

File: tests/testthat/test_xgboost_1.R (hunk 15 above)
Comment (hideaki) [reply tests/testthat/test_xgboost_1.R:204]: Done.

File: tests/testthat/test_xgboost_1.R (hunk 16 above)
Comment (hideaki) [reply tests/testthat/test_xgboost_1.R:428]: Done.

**Pull request merged**